File: src/main/java/org/o3project/odenos/core/component/ConversionTable.java
/*
 * Copyright 2015 NEC Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.o3project.odenos.core.component;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/**
 * ConversionTable class.
 *
 */
public class ConversionTable {

  /**
   * Constructor.
   */
  public ConversionTable() {
  }

  // ex
  // key : connectionId
  // value : connectionType
  // {connA:original, connB:sliver, connC:sliver}
  private HashMap<String, String> connectionTypeMap = new HashMap<String, String>();

  // key : connectionType
  // value : connectionIdList
  // {original:[connA], sliver:[connB,CoonC]}
  // private HashMap<String, ArrayList<String>>
  // connectionTypeTable = new HashMap<String, ArrayList<String>>();

  /**
   * Returns a type of connection.
   * @param connectionId connection ID.
   * @return type of connection.
   */
  public final String getConnectionType(final String connectionId) {
    return connectionTypeMap.get(connectionId);
  }

  /**
   * Returns a list of connection IDs.
   * @param connectionType type of connection.
   * @return list of connection IDs.
   */
  public final ArrayList<String>
      getConnectionList(final String connectionType) {
    ArrayList<String> connectionIds = new ArrayList<String>();
    for (Map.Entry<String, String> e : connectionTypeMap.entrySet()) {
      if (connectionType.equals(e.getValue())) {
        connectionIds.add(e.getKey());
      }
    }
    return connectionIds;
  }

  /**
   * Returns true if this contained type.
   * @param connectionType type of
   * @return true if this contained type.
   */
  public final boolean isConnectionType(final String connectionType) {
    if (connectionType == null
        || 0 == this.getConnectionList(connectionType).size()) {
      return false;
    }
    return true;
  }

  /**
   * Associates the type in connection map.
   * @param id key.
   * @param type value of type to be associated with the key.
   */
  public final void addEntryConnectionType(
      final String id,
      final String type) {
    connectionTypeMap.put(id, type);
  }

  /**
   * Removes the mapping for a key from connection map.
   * @param id key.
   */
  public final void delEntryConnectionType(final String id) {
    connectionTypeMap.remove(id);
  }

  // ex
  // connA --- [LogicComponent] --+-- connB
  // +-- connC
  //
  // connA : [connB,connC]
  // connB : [connA]
  // connC : [connA]

  private HashMap<String, ArrayList<String>> nwcConversionTable =
      new HashMap<String, ArrayList<String>>();
  private HashMap<String, ArrayList<String>> nodeConversionTable =
      new HashMap<String, ArrayList<String>>();
  private HashMap<String, ArrayList<String>> portConversionTable =
      new HashMap<String, ArrayList<String>>();
  private HashMap<String, ArrayList<String>> linkConversionTable =
      new HashMap<String, ArrayList<String>>();
  private HashMap<String, ArrayList<String>> flowConversionTable =
      new HashMap<String, ArrayList<String>>();

  // get object connection
  /**
   * Returns a conversion map of network.
   * @return conversion map of network.
   */
  public final HashMap<String, ArrayList<String>> getNetwork() {
    return nwcConversionTable;
  }

  /**
   * Returns a list of network IDs.
   * @param nwcId network ID.
   * @return list of network IDs.
   */
  public final ArrayList<String> getNetwork(final String nwcId) {
    ArrayList<String> ret = nwcConversionTable.get(nwcId);
    if (ret == null) {
      return new ArrayList<String>();
    }
    return ret;
  }

  /**
   * Returns a conversion map of node.
   * @return conversion map of node.
   */
  public final HashMap<String, ArrayList<String>> getNode() {
    return nodeConversionTable;
  }

  /**
   * Returns a list of node IDs.
   * @param nwcId network ID.
   * @param nodeId node ID.
   * @return list of node IDs.
   */
  public final ArrayList<String> getNode(final String nwcId, final String nodeId) {
    String key = nwcId + "::" + nodeId;
    if (nodeConversionTable.containsKey(key)) {
      return nodeConversionTable.get(key);
    }
    return new ArrayList<String>();
  }

  /**
   * Returns a conversion map of port.
   * @return conversion map of port.
   */
  public final HashMap<String, ArrayList<String>> getPort() {
    return portConversionTable;
  }

  /**
   * Returns a list of port IDs.
   * @param nwcId network ID.
   * @param nodeId node ID.
   * @param portId port ID.
   * @return list of port IDs.
   */
  public final ArrayList<String>
      getPort(final String nwcId, final String nodeId, final String portId) {
    String key = nwcId + "::" + nodeId + "::" + portId;
    if (portConversionTable.containsKey(key)) {
      return portConversionTable.get(key);
    }
    return new ArrayList<String>();
  }

  /**
   * Returns a conversion map of link.
   * @return conversion map of link.
   */
  public final HashMap<String, ArrayList<String>> getLink() {
    return linkConversionTable;
  }

  /**
   * Returns a list of link IDs.
   * @param nwcId network ID.
   * @param linkId link ID.
   * @return list of link IDs.
   */
  public final ArrayList<String>
      getLink(final String nwcId, final String linkId) {
    String key = nwcId + "::" + linkId;
    if (linkConversionTable.containsKey(key)) {
      return linkConversionTable.get(key);
    }
    return new ArrayList<String>();
  }

  /**
   * Returns a conversion map of flow.
   * @return conversion map of flow.
   */
  public final HashMap<String, ArrayList<String>> getFlow() {
    return flowConversionTable;
  }

  /**
   * Returns a list of flow IDs.
   * @param nwcId network ID.
   * @param flowId flow ID.
   * @return list of flow IDs.
   */
  public final ArrayList<String>
      getFlow(final String nwcId, final String flowId) {
    String key = nwcId + "::" + flowId;
    if (flowConversionTable.containsKey(key)) {
      return flowConversionTable.get(key);
    }
    return new ArrayList<String>();
  }

  // add object connection
  /**
   * Associates the network in network map.
   * @param nwcId1 network ID as key.
   * @param nwcId2 associated network ID.
   */
  public final void addEntryNetwork(
      final String nwcId1,
      final String nwcId2) {
    addEntryObject(nwcConversionTable, nwcId1, nwcId2);
  }

  /**
   * Appends the node to this table.
   * @param orgNwcId original network ID.
   * @param orgNodeId original node ID.
   * @param repNwcId replaced network ID.
   * @param repNodeId replaced node ID.
   */
  public final void addEntryNode(
      final String orgNwcId,
      final String orgNodeId,
      final String repNwcId,
      final String repNodeId) {
    String key = orgNwcId + "::" + orgNodeId;
    String value = repNwcId + "::" + repNodeId;
    addEntryObject(nodeConversionTable, key, value);
  }

  /**
   * Appends the port to this table.
   * @param orgNwcId original network ID.
   * @param orgNodeId original node ID.
   * @param orgPortId original port ID.
   * @param repNwcId replaced network ID.
   * @param repNodeId replaced node ID.
   * @param repPortId replaced port ID.
   */
  public final void addEntryPort(
      final String orgNwcId,
      final String orgNodeId,
      final String orgPortId,
      final String repNwcId,
      final String repNodeId,
      final String repPortId) {
    String key = orgNwcId + "::" + orgNodeId + "::" + orgPortId;
    String value = repNwcId + "::" + repNodeId + "::" + repPortId;
    addEntryObject(portConversionTable, key, value);
  }

  /**
   * Appends the link to this table.
   * @param orgNwcId original network ID.
   * @param orgLinkId original link ID.
   * @param repNwcId replaced network ID.
   * @param repLinkId replaced link ID.
   */
  public final void addEntryLink(
      final String orgNwcId,
      final String orgLinkId,
      final String repNwcId,
      final String repLinkId) {
    String key = orgNwcId + "::" + orgLinkId;
    String value = repNwcId + "::" + repLinkId;
    addEntryObject(linkConversionTable, key, value);
  }

  /**
   * Appends the flow to this table.
   * @param orgNwcId original network ID.
   * @param orgFlowId original flow ID.
   * @param repNwcId replaced network ID.
   * @param repFlowId replaced flow ID.
   */
  public final void addEntryFlow(
      final String orgNwcId,
      final String orgFlowId,
      final String repNwcId,
      final String repFlowId) {
    String key = orgNwcId + "::" + orgFlowId;
    String value = repNwcId + "::" + repFlowId;
    addEntryObject(flowConversionTable, key, value);
  }

  private void addEntryObject(
      final HashMap<String, ArrayList<String>> hashObj,
      final String key,
      final String value) {

    // key setting
    ArrayList<String> valueList = hashObj.get(key);
    if (valueList == null) {
      valueList = new ArrayList<String>();
      hashObj.put(key, valueList);
    }
    valueList.add(value);

    // value -> key setting(reverse setting)
    ArrayList<String> valueListRev = hashObj.get(value);
    if (valueListRev == null) {
      valueListRev = new ArrayList<String>();
      hashObj.put(value, valueListRev);
    }
    valueListRev.add(key);
  }

  // delete object connection
  /**
   * Removes the mapping for a key from network map.
   * @param key network ID as key.
   */
  public final void delEntryNetwork(
      final String key) {
    delEntryObject(nwcConversionTable, key);
    return;
  }

  /**
   * Delete the node ID from in this table.
   * @param nwcId network ID.
   * @param nodeId node ID.
   */
  public final void delEntryNode(
      final String nwcId,
      final String nodeId) {
    // Delete Port => Node.
    ArrayList<String> delPortList = new ArrayList<String>();
    for (String pid : portConversionTable.keySet()) {
      String[] plist = pid.split("::");
      if (plist[0].equals(nwcId)
          && plist[1].equals(nodeId)) {
        delPortList.add(pid);
      }
    }
    for (String pid : delPortList) {
      delEntryObject(portConversionTable, pid);

    }
    String key = nwcId + "::" + nodeId;
    delEntryObject(nodeConversionTable, key);
    return;
  }

  /**
   * Delete the port ID from in this table.
   * @param nwcId network ID.
   * @param nodeId node ID.
   * @param portId port ID.
   */
  public final void delEntryPort(
      final String nwcId,
      final String nodeId,
      final String portId) {
    String key = nwcId + "::" + nodeId + "::" + portId;
    delEntryObject(portConversionTable, key);
    return;
  }

  /**
   * Delete the link ID from in this table.
   * @param nwcId network ID.
   * @param linkId link ID.
   */
  public final void delEntryLink(
      final String nwcId,
      final String linkId) {
    String key = nwcId + "::" + linkId;
    delEntryObject(linkConversionTable, key);
    return;
  }

  /**
   * Delete the flow ID from in this table.
   * @param nwcId network ID.
   * @param flowId flow ID.
   */
  public final void delEntryFlow(
      final String nwcId,
      final String flowId) {
    String key = nwcId + "::" + flowId;
    delEntryObject(flowConversionTable, key);
    return;
  }

  private void delEntryObject(
      final HashMap<String, ArrayList<String>> hashObj,
      final String key) {
    ArrayList<String> valueList = hashObj.get(key);
    if (valueList == null) {
      return;
    }

    // value -> key remove(reverse setting remove)
    for (String reversekey : valueList) {
      if (!hashObj.containsKey(reversekey)) {
        continue;
      }
      if (hashObj.get(reversekey).size() > 1) {
        hashObj.get(reversekey).remove(key);
        continue;
      }
      hashObj.remove(reversekey);
    }
    hashObj.remove(key);
  }
}

File: src/main/python/org/o3project/odenos/core/component/conversion_table.py
# -*- coding:utf-8 -*-

# Copyright 2015 NEC Corporation.                                          #
#                                                                          #
# Licensed under the Apache License, Version 2.0 (the "License");          #
# you may not use this file except in compliance with the License.         #
# You may obtain a copy of the License at                                  #
#                                                                          #
#   http://www.apache.org/licenses/LICENSE-2.0                             #
#                                                                          #
# Unless required by applicable law or agreed to in writing, software      #
# distributed under the License is distributed on an "AS IS" BASIS,        #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. #
# See the License for the specific language governing permissions and      #
# limitations under the License.                                           #


class ConversionTable(object):

    def __init__(self):
        self.__connection_type_map = {}

        self.__network_conversion_table = {}
        self.__node_conversion_table = {}
        self.__port_conversion_table = {}
        self.__link_conversion_table = {}
        self.__flow_conversion_table = {}

    def get_connection_type(self, connection_id):
        if connection_id in self.__connection_type_map:
            return self.__connection_type_map[connection_id]
        return None

    def get_connection_list(self, connection_type):
        connection_ids = []
        for k_conn_id, v_conn_type in self.__connection_type_map.items():
            if connection_type == v_conn_type:
                connection_ids.append(k_conn_id)

        return connection_ids

    def is_connection_type(self, connection_type):
        if connection_type is None or\
           len(self.get_connection_list(connection_type)) == 0:
            return False

        return True

    def add_entry_connection_type(self, connection_id, connection_type):
        self.__connection_type_map[connection_id] = connection_type

    def del_entry_connection_type(self, connection_id):
        if self.__connection_type_map.has_key(connection_id):
            del self.__connection_type_map[connection_id]

    def get_network(self, network_id):
        networks = []
        if network_id in self.__network_conversion_table:
            networks = self.__network_conversion_table[network_id]

        return networks

    def get_node(self, network_id, node_id):
        nodes = []
        key = network_id + "::" + node_id
        if key in self.__node_conversion_table:
            nodes = self.__node_conversion_table[key]

        return nodes

    def get_port(self, network_id, node_id, port_id):
        ports = []
        key = network_id + "::" + node_id + "::" + port_id
        if key in self.__port_conversion_table:
            ports = self.__port_conversion_table[key]

        return ports

    def get_link(self, network_id, link_id):
        links = []
        key = network_id + "::" + link_id
        if key in self.__link_conversion_table:
            links = self.__link_conversion_table[key]

        return links

    def get_flow(self, network_id, flow_id):
        flows = []
        key = network_id + "::" + flow_id
        if key in self.__flow_conversion_table:
            flows = self.__flow_conversion_table[key]

        return flows

    def add_entry_network(self, nwc_id_1, nwc_id_2):
        self.__add_entry_object(self.__network_conversion_table,
                                nwc_id_1,
                                nwc_id_2)

    def add_entry_node(self, org_nwc_id, org_node_id,
                       rep_nwc_id, rep_node_id):
        key = org_nwc_id + "::" + org_node_id
        value = rep_nwc_id + "::" + rep_node_id
        self.__add_entry_object(self.__node_conversion_table,
                                key,
                                value)

    def add_entry_port(self, org_nwc_id, org_node_id, org_port_id,
                       rep_nwc_id, rep_node_id, rep_port_id):
        key = org_nwc_id + "::" + org_node_id + "::" + org_port_id
        value = rep_nwc_id + "::" + rep_node_id + "::" + rep_port_id
        self.__add_entry_object(self.__port_conversion_table,
                                key,
                                value)

    def add_entry_link(self, org_nwc_id, org_link_id,
                       rep_nwc_id, rep_link_id):
        key = org_nwc_id + "::" + org_link_id
        value = rep_nwc_id + "::" + rep_link_id
        self.__add_entry_object(self.__link_conversion_table,
                                key,
                                value)

    def add_entry_flow(self, org_nwc_id, org_flow_id,
                       rep_nwc_id, rep_flow_id):
        key = org_nwc_id + "::" + org_flow_id
        value = rep_nwc_id + "::" + rep_flow_id
        self.__add_entry_object(self.__flow_conversion_table,
                                key,
                                value)

    def __add_entry_object(self, conv_table_obj, key, value):
        # key setting
        if key not in conv_table_obj:
            conv_table_obj[key] = []
        conv_table_obj[key].append(value)

        # value -> key setting(reverse setting)
        if value not in conv_table_obj:
            conv_table_obj[value] = []
        conv_table_obj[value].append(key)

    def del_entry_network(self, key):
        self.__del_entry_object(self.__network_conversion_table, key)

    def del_entry_node(self, network_id, node_id):
        # delete Port => Node.
        del_port_list = []
        for port_id in self.__port_conversion_table:
            port_list = port_id.split("::")
            if port_list[0] == network_id and\
               port_list[1] == node_id:
                del_port_list.append(port_id)

        for port_id in del_port_list:
            self.__del_entry_object(self.__port_conversion_table,
                                    port_id)

        key = network_id + "::" + node_id
        self.__del_entry_object(self.__node_conversion_table, key)

    def del_entry_port(self, network_id, node_id, port_id):
        key = network_id + "::" + node_id + "::" + port_id
        self.__del_entry_object(self.__port_conversion_table, key)

    def del_entry_link(self, network_id, link_id):
        key = network_id + "::" + link_id
        self.__del_entry_object(self.__link_conversion_table, key)

    def del_entry_flow(self, network_id, flow_id):
        key = network_id + "::" + flow_id
        self.__del_entry_object(self.__flow_conversion_table, key)

    def __del_entry_object(self, conv_table_obj, key):
        if key not in conv_table_obj:
            return

        # value -> key remove(reverse setting remove)
        reverse_keys = conv_table_obj[key]
        for reverse_key in reverse_keys:
            if reverse_key not in conv_table_obj:
                continue
            if len(conv_table_obj[reverse_key]) > 1:
                conv_table_obj[reverse_key].remove(key)
                continue

            del conv_table_obj[reverse_key]

        del conv_table_obj[key]
